### DISCUSSION TITLE
Signpost closing doesn't respect two way binding

### DISCUSSION THREAD

**Discussion by towilliamstwilio** [vmware-clarity/ng-clarity#801] [General]
Hi - curious if anyone has experienced this but it seems signposts no longer respect two way binding. It works to open the signpost, but it doesn't work to close it. HTML is below:

```
<clr-signpost id="{{ip['IP']}}" (click)="expandIPInfo($event.currentTarget.id)">
                                  <clr-signpost-content *clrIfOpen="IPInfo">
                                      <h3 style="margin-top: 0">{{ip['IP']}}</h3>
                                      <div *ngIf="checkIMIPResponse;"><button class="btn btn-sm btn-link" (click)="addIMIPModal($event)">Add to List</button></div>
                                      <button class="btn btn-sm btn-link" (click)="enableDeleteIPModal($event)">Delete</button>
                                      <br />
                                      <span>
                                        Datacoming soon</span>                        
                                  </clr-signpost-content>
                              </clr-signpost>
```

  
Clicking on the signpost icon triggers expandIPInfo, which sets 'IPInfo = true' and opens the signpost. But when clicking on Delete within the signpost, it triggers enableDeleteIPModal, which sets 'IPInfo = false' and should close the signpost. But that's not happening. The signpost remains open. TS for enableDeleteIpModal is below:

`public enableDeleteIPModal (event: any) {
  this.IPInfo = false;
  this.deletingIP = true;
}`                            

This worked on previous versions of Angular and Clarity. I recently upgraded everything and now I'm encountering this issue. 

Node: 18.16
Angular CLI: 14.2.12
Angular 14.3.0
Clarity: 13.18.2

**Comment by towilliamstwilio** [comment]
I also tried using *ngIf and the same result. The signpost only closes if you click somewhere else on the screen or if you click the x on the signpost.

**Comment by kevinbuhmann** [comment]
The "Delete" button click event is bubbling up to the `clr-signpost` click event handler causing the signpost to be re-opened.

You can fix this by 1) calling `event.stopPropagation()` in the "Delete" button click event handler or 2) using two-way binding.

Your code example is using one-way binding. It does not handle the `(clrIfOpenChange)` output to update the `IPInfo` property in your component when the signpost is opened. This is required so that Angular can detect the change when it is set to false in the `enableDeleteIPModal` method.

The `*directive` syntax does not support handling events from the structural directive; however, you can use the de-sugared `ng-template` syntax to handle the `(clrIfOpenChange)` output or use the `[(clrIfOpen)]` two-way binding syntax (which is just syntactic sugar for an input and output binding in a single binding).

See here for an example of how to implement this: https://stackblitz.com/edit/clarity-light-theme-clr13-cds6-nwa2ss?file=src%2Fapp%2Fapp.component.html

**Comment by towilliamstwilio** [comment]
Thanks, Kevin! Neither of these options seem to be working. It might be because I'm nesting the signposts in list items inside a card? I downgraded Node just to be safe since I was on 18. I'm running 16 now and still the same issues.

`ng-template` results in all signposts in the list to open. Updated code and screenshot below. Additionally, the functionality behind the delete and add to list buttons no longer fires. Clicking the buttons closes the signposts, but the functionality doesn't work beyond that.

```
<div class="card-block">
                              <h3 class="card-title"><strong>IP Addresses</strong></h3>
                              <ul class="list" *ngFor="let ip of convertIPsResponse">
                                <li>{{ip['IP']}}
                                <clr-signpost id="{{ip['IP']}}" (click)="expandIPInfo($event.currentTarget.id); $event.stopPropagation()">
                                  <ng-template [(clrIfOpen)]="IPInfo">
                                  <clr-signpost-content>
                                      <h3 style="margin-top: 0">{{ip['IP']}}</h3>
                                      <div *ngIf="checkIMIPResponse"><button class="btn btn-sm btn-link" (click)="addIMIPModal($event); $event.stopPropagation()">Add to List</button></div>
                                      <button class="btn btn-sm btn-link" (click)="enableDeleteIPModal($event); $event.stopPropagation()">Delete</button>
                                      <br />
                                      <span>
                                        Data coming soon</span>
                                      </clr-signpost-content>
                                  </ng-template>
                              </clr-signpost>
                              </li>
                              </ul>
                          </div>
```
<img width="548" alt="Screen Shot 2023-07-06 at 4 34 34 PM" src="https://github.com/vmware-clarity/ng-clarity/assets/98768436/7ef8411d-2873-4a54-9c9d-243d19c4977f">

And if I use `*clrIfOpen` method with `event.stopPropagation`, the delete and add buttons work but the signpost doesn't close. Code and screenshot below:

```
<div class="card-block">
                              <h3 class="card-title"><strong>IP Addresses</strong></h3>
                              <ul class="list" *ngFor="let ip of convertIPsResponse">
                                <li>{{ip['IP']}}
                                <clr-signpost id="{{ip['IP']}}" (click)="expandIPInfo($event.currentTarget.id)">
                                  <clr-signpost-content *clrIfOpen="IPInfo">
                                      <h3 style="margin-top: 0">{{ip['IP']}}</h3>
                                      <div *ngIf="checkIMIPResponse"><button class="btn btn-sm btn-link" (click)="addIMIPModal($event); $event.stopPropagation()">Add to List</button></div>
                                      <button class="btn btn-sm btn-link" (click)="enableDeleteIPModal($event); $event.stopPropagation()">Delete</button>
                                      <br />
                                      <span>
                                        Data coming soon</span>
                                      </clr-signpost-content>
                              </clr-signpost>
                              </li>
                              </ul>
                          </div>

public enableDeleteIPModal (event: any) {
  this.IPInfo = false;
  this.deletingIP = true;
}
```

<img width="1136" alt="Screen Shot 2023-07-06 at 4 44 14 PM" src="https://github.com/vmware-clarity/ng-clarity/assets/98768436/9126f0b1-44b1-4d33-9de1-ebfa5adc7b7e">

Now if I use `*ngIf ` with `event.stopPropagation` instead of `*clrIfOpen` then the signpost closes as expected, but the content in the signpost is messed up. For instance, content for 1.2.3.4 appears in the signpost for the first item in the list.

**Reply by kevinbuhmann** [reply]
>It might be because I'm nesting the signposts in list items inside a card? [...] `ng-template` results in all signposts in the list to open.

You will need a different property for each signpost. If you bind to the same value for all of the signposts, that will happen.

>Additionally, the functionality behind the delete and add to list buttons no longer fires. Clicking the buttons closes the signposts, but the functionality doesn't work beyond that.

Please create reproduction via StackBlitz. https://stackblitz.com/@clr-team

**Reply by kevinbuhmann** [reply]
Also, note that the signpost is NOT intended for interactive content. What you are trying to do should work technically, but it is not recommended to have buttons in the signposts. I would explore an alternative UX before investing too much time on this because this approach will very likely have accessibility issues. The signpost is for lightweight contextual content. It does not trap focus or do anything to else to support interactive content.

>With the exception of text/image links, don't insert any actionable components, such as buttons, in the dialog.
-- https://clarity.design/documentation/signposts
